### ISSUE TITLE
Special-case __debug__ in assignments etc.

### ISSUE THREAD

**Issue by gvanrossum** [we-like-parsers/cpython#109]
Basically we should duplicate all the forbidden_name() calls from ast.c (but only checking for `__debug__`).

**Comment by gvanrossum** [comment]
Some cases that still need to be addressed (I guess there are no tests):
```
__debug__: int
class C(__debug__=1): pass
def f(__debug__): pass
f(__debug__=42)
```
My most recent commit takes care of most others:
- plain and attribute assignments
- imports
- function/class names
- except clause

**Comment by gvanrossum** [comment]
So I found that there's a bigger problem, star_target is too general. E.g.
```
(a, *True) = 1, 1
```
causes a segfault (the same as before in #98, though we got there in a different way).

The problem is that the grammar currently allows `star_target: '*' bitwise_or`. I have a tentative fix but am not sure whether it will work:
```
diff --git a/Grammar/python.gram b/Grammar/python.gram
index 4f6b1c5290..e9f76832d1 100644
--- a/Grammar/python.gram
+++ b/Grammar/python.gram
@@ -472,7 +472,8 @@ star_targets[expr_ty]:
         _Py_Tuple(CHECK(_PyPegen_seq_insert_in_front(p, a, b)), Store, EXTRA) }
 star_targets_seq[asdl_seq*]: a=','.star_target+ [','] { a }
 star_target[expr_ty] (memo):
-    | '*' a=bitwise_or { _Py_Starred(CHECK(_PyPegen_set_expr_context(p, a, Store)), Store, EXTRA) }
+    | '*' a=(star_atom | t_primary | invalid_target) {
+        _Py_Starred(CHECK(_PyPegen_set_expr_context(p, a, Store)), Store, EXTRA) }
     | a=t_primary '.' b=NAME !t_lookahead { _Py_Attribute(a, b->v.Name.id, Store, EXTRA) }
     | a=t_primary '[' b=slices ']' !t_lookahead { _Py_Subscript(a, b, Store, EXTRA) }
     | star_atom
@@ -538,9 +539,11 @@ invalid_assignment:
     | expression ':' expression ['=' annotated_rhs] {
         RAISE_SYNTAX_ERROR("illegal target for annotation") }
     | a=expression ('=' | augassign) (yield_expr | star_expressions) {
-        RAISE_SYNTAX_ERROR("cannot assign to %s", _PyPegen_get_expr_name(a)) }
+        RAISE_SYNTAX_ERROR("AAA cannot assign to %s", _PyPegen_get_expr_name(a)) }
 invalid_block:
     | NEWLINE !INDENT { RAISE_INDENTATION_ERROR("expected an indented block") }
 invalid_comprehension:
     | ('[' | '(' | '{') '*' expression for_if_clauses {
         RAISE_SYNTAX_ERROR("iterable unpacking cannot be used in comprehension") }
+invalid_target:
+    | a=bitwise_or { RAISE_SYNTAX_ERROR("BBB cannot assign to %s", _PyPegen_get_expr_name(a)) }
```

This would have to be tested against all of the stdlib, at least, to validate the idea.

The errors still aren't perfect.

**Comment by gvanrossum** [comment]
Well it looks like `make time_stdlib` in Tools/peg_generator works, so I am less worried. But I'll leave it up to you to land it, since I won't have time to work more on anything today.

**Comment by gvanrossum** [comment]
Fix pushed as e2bd5f64514c45d19e3589c28fda32d84ad977e6 -- but I'm leaving it up to you to deal with any fallout (though I am not aware of any).

**Comment by gvanrossum** [comment]
So assignments to `__debug__` in various disguises are now done. The one difference is that it happens during bytecode compilation rather than during AST construction.

Separately, `(a, *True) = ...` produces the message "cannot assign to tuple" rather than "cannot assign to True". I'll file a new issue for that.
